cluster/tls_connection.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/binary"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/hashicorp/memberlist"
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+)
+
+const (
+	version      = "v1.0.0"
+	uint32length = 4
+)
+
+// tlsConn wraps net.Conn with connection pooling data.
+type tlsConn struct {
+	connection net.Conn
+	timeout    time.Duration
+	lock       sync.Mutex
+	live       bool
+}
+
+func dialTLSConn(addr string, timeout time.Duration, tlsConfig *tls.Config) (*tlsConn, error) {
+	dialer := &net.Dialer{Timeout: timeout}
+	conn, err := tls.DialWithDialer(dialer, network, addr, tlsConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tlsConn{
+		connection: conn,
+		timeout:    timeout,
+		live:       true,
+	}, nil
+}
+
+func rcvTLSConn(conn net.Conn) *tlsConn {
+	return &tlsConn{
+		connection: conn,
+		live:       true,
+	}
+}
+
+// Write writes a byte array into the connection. It returns the number of bytes written and an error.
+func (conn *tlsConn) Write(b []byte) (int, error) {
+	conn.lock.Lock()
+	defer conn.lock.Unlock()
+	n, err := conn.connection.Write(b)
+
+	if err != nil {
+		conn.live = false
+	}
+	return n, err
+}
+
+func (conn *tlsConn) alive() bool {
+	conn.lock.Lock()
+	defer conn.lock.Unlock()
+	return conn.live
+}
+
+// writePacket writes all the bytes in one operation so no concurrent write happens in between.
+// It prefixes the message length.
+func (conn *tlsConn) writePacket(fromAddr string, b []byte) error {
+	msg, err := proto.Marshal(&clusterpb.MemberlistMessage{
+		Version:  version,
+		Kind:     clusterpb.MemberlistMessage_PACKET,
+		FromAddr: fromAddr,
+		Msg:      b,
+	})
+	if err != nil {
+		return errors.Wrap(err, "unable to marshal memeberlist packet message")
+	}
+	buf := make([]byte, uint32length)
```

Comment:
Probably not a huge deal, but by making this buffer only 4 bytes an extra allocation is happening for each write. Instead this buffer should probably be len(msg) + 4 bytes long, then you can PutUint32 and copy msg in without an extra allocation.

Same with writeStream()

Reply:
Thanks, Chris. Does adding a capacity argument solve this issue? Like so https://github.com/prometheus/alertmanager/pull/2237/commits/1018b286c8dd998ff74d6ce7c9b843a6f651b8e0 ?